datafusion/sql/src/unparser/plan.rs

```diff
@@ -476,22 +474,72 @@ impl Unparser<'_> {
                 self.select_to_sql_recursively(input, query, select, relation)
             }
             LogicalPlan::Join(join) => {
-                let join_constraint = self.join_constraint_to_sql(
-                    join.join_constraint,
-                    &join.on,
-                    join.filter.as_ref(),
+
+                let mut table_scan_filters = vec![];
+
+                let left_plan = match try_transform_to_simple_table_scan_with_filters(&join.left) {
+                    Some((plan, filters)) => {
+                        table_scan_filters.extend(filters);
+                        Arc::new(plan)
+                    }
+                    None => Arc::clone(&join.left)
+                };
+
+                self.select_to_sql_recursively(
+                    left_plan.as_ref(),
+                    query,
+                    select,
+                    relation,
                 )?;
 
+                let right_plan = match try_transform_to_simple_table_scan_with_filters(&join.right) {
+                    Some((plan, filters)) => {
+                        table_scan_filters.extend(filters);
+                        Arc::new(plan)
+                    }
+                    None => Arc::clone(&join.right)
+                };
+
                 let mut right_relation = RelationBuilder::default();
 
                 self.select_to_sql_recursively(
-                    join.left.as_ref(),
+                    right_plan.as_ref(),
                     query,
                     select,
-                    relation,
+                    &mut right_relation,
                 )?;
+
+                let join_filters = if table_scan_filters.is_empty() {
+                    join.filter.clone()
+                } else {
+                    // Combine `table_scan_filters` into a single filter using `AND`
+                    let combined_filters = table_scan_filters
+                        .into_iter()
+                        .reduce(|acc, filter| Expr::BinaryExpr(BinaryExpr {
+                            left: Box::new(acc),
+                            op: Operator::And,
+                            right: Box::new(filter),
+                        }))
+                        .unwrap();
```

Comment:
Can this at least be an `.expect` with a message on why it shouldn't panic?

Reply:
Fixed - thank you